### PR TITLE
Introduce a new tensor type for return_tensors on tokenizer for NumPy

### PR DESCRIPTION
**Pull request by mfuntowicz**
Two changes in this PR:

- As we're introducing more than two tensor backend alternatives I created an enum `TensorType` listing all the possible tensor we can create `TensorType.TENSORFLOW`, `TensorType.PYTORCH`, `TensorType.NUMPY`. This might help newcomers who don't know about `"tf"`, `"pt"`.

   _->Note: TensorType are compatible with previous  `"tf"`, `"pt"` and now `"np"` str to allow backward compatbility (+unittest)_

- Numpy is now a possible target when creating tensors. This is usefull for JAX :)

### CODE REVIEW

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -58,6 +72,12 @@
 EncodedInputPair = Tuple[List[int], List[int]]
 
 
+class TensorType(Enum):
+    PYTORCH = "pt"
+    TENSORFLOW = "tf"
+    NUMPY = "np"
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py:78]: maybe list should be in here?

File: src/transformers/tokenization_utils.py [hunk 2]
```diff
@@ -161,6 +181,46 @@ def truncate_and_pad(
...
+
+            # at-least2d
+            if tensor.ndim > 2:
+                tensor.squeeze_(0)
+            elif tensor.ndim < 2:
+                tensor = tensor[None, :]
+
+            batch_outputs[key] = tensor
+        except ValueError:
+            if None in [item for sequence in value for item in sequence]:
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py:216]: how do we know that `None` is padding issue?

**Review by sshleifer — APPROVED** [review]
LGTM

File: src/transformers/tokenization_utils.py [hunk 3]
```diff
@@ -161,6 +181,46 @@ def truncate_and_pad(
...
+    # Do the tensor conversion in batch
+    for key, value in batch_outputs.items():
+        try:
+            tensor = as_tensor(value)
+
+            # at-least2d
+            if tensor.ndim > 2:
+                tensor.squeeze_(0)
+            elif tensor.ndim < 2:
+                tensor = tensor[None, :]
```
Comment (LysandreJik) [inline src/transformers/tokenization_utils.py:212]: Cool unsqueeze trick!

**Review by LysandreJik — APPROVED** [review]
LGTM! I think we can increase the test coverage here, by testing that it correctly raises/correctly outputs what we're looking for. We already have the test `test_batch_encode_plus_tensors` and `test_torch_encode_plus_sent_to_model`/`test_tf_encode_plus_sent_to_model`. Do you think we could test NumPy as well?

File: src/transformers/tokenization_utils.py (hunk 2 above)
Comment (mfuntowicz) [reply src/transformers/tokenization_utils.py:216]: I suspect it comes from the fact that if padding is not set, then the actual `padding`/`padding_idx` are `None`. 

This snippet didn't change with current master, so I didn't take too much time investigating here tbh. It may make sense to dig further if we actually get None in padded input (@LysandreJik any input on this?)

**Comment by mfuntowicz** [comment]
@LysandreJik I pushed a initial version of `test_np_encode_plus_sent_to_model` which converts input to numpy tensor. 

For the moment we don't have any model to forward through (JAX/Flax PR is not merged). I added a note to complete the unittests when we have the full pipeline available.

**Review by thomwolf — APPROVED** [review]
Very nice!
